domain-models-runtime/src/main/java/org/folio/rest/persist/PostgresConnectionManager.java

```diff
@@ -0,0 +1,128 @@
+package org.folio.rest.persist;
+
+import io.vertx.pgclient.PgConnection;
+import io.vertx.sqlclient.Pool;
+import io.vertx.core.Future;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.Optional;
+
+public class PostgresConnectionManager {
+  private static final Logger LOG = LogManager.getLogger(PostgresConnectionManager.class);
+  private final Collection<CachedPgConnection> connectionCache = Collections.synchronizedCollection(new ArrayList<>());
+
+  private int cacheHits;
+  private int cacheMisses;
+  private int activeConnectionCount;
+
+  public PostgresConnectionManager() {
+    // TODO
+  }
+
+  public int getCacheSize() {
+    return connectionCache.size();
+  }
+
+  public void clearCache() {
+    cacheHits = 0;
+    cacheMisses = 0;
+    activeConnectionCount = 0;
+    connectionCache.clear();
+    LOG.debug("Cleared connection manager");
+  }
+
+  public void tryClose(CachedPgConnection connection) {
+    this.tryAddToCache(connection);
+    this.tryRemoveOldestAvailableConnectionAndClose();
+  }
+
+  public Future<PgConnection> getConnection(Pool pool, String schemaName, String tenantId) {
+    if (! PostgresClient.sharedPgPool) {
+      LOG.debug("Not in shared pool mode");
+      return pool.getConnection().map(PgConnection.class::cast);
+    }
+
+    LOG.debug("In shared pool mode");
+    return getOrCreateCachedConnection(pool, schemaName, tenantId);
+  }
+
+  private void tryAddToCache(CachedPgConnection connection) {
+    if (connectionCache.contains(connection)) {
+      LOG.debug("Item already exists in cache: {} {} {}",
+          connection.getTenantId(), connection.getSessionId(), connection.isAvailable());
+      return;
+    }
+    connectionCache.add(connection);
+  }
+
+  private void tryRemoveOldestAvailableConnectionAndClose() {
+    // The cache must not grow larger than the max pool size of the underlying pool.
+    // TODO get the pool size from the env.
+    var cacheExhausted = this.connectionCache.size() >= PostgresClient.DEFAULT_MAX_POOL_SIZE;
+    if (! cacheExhausted) {
+      LOG.debug("Cache is not yet exhausted");
+      return;
+    }
+
+    connectionCache.stream()
+        .filter(CachedPgConnection::isAvailable)
+        .min(Comparator.comparingLong(CachedPgConnection::getTimestamp))
+        .ifPresent(connection -> {
+          LOG.debug("Removing and closing oldest available connection: {} {}",
+              connection.getTenantId(), connection.getSessionId());
+          if (connection.getWrappedConnection() != null) {
+            activeConnectionCount--;
+            connection.getWrappedConnection().close();
+          }
+          connectionCache.remove(connection);
+        });
+  }
```

Comment:
Ok agree we need to handle timeout for idle connections since even though the connections will cycle through fairly quickly under load there could be connections that hang around during idle times.

Reply:
I took a stab at implementing a timer tied to the the `Vertx` of `PostgresClient`. Since there can be multiple clients with different configs, on consideration was to let the lowest connection release delay be the one controlling eviction based on the timeout.

Reply:
I think I am not yet configuring the idleTimeout of PgPool to be 0. Let me do that.

Reply:
I think I have decided to go ahead with the idea of making the timeout a feature of the overridden connection class. This will be a lot cleaner. Also I think it could be that the timeouts in the PTF tests are the result of idleTimeout not being set to 0 (fully overriding the internal timeout). If we fully override we fully control it.

Reply:
Please see the latest for a new implementation of the connection release delay (the [ReleaseDelayObserver](https://github.com/folio-org/raml-module-builder/pull/1142/files#diff-df1fdffb5f8379246f2a6659389f2a203b589a51cabbf2e757a12d1f3b056884)). This allows each connection to manage its own release delay timer which is reset each time a connection becomes available.